test/cloudwatchlogs/publish_logs_test.go

```diff
@@ -113,6 +108,104 @@ func TestWriteLogsToCloudWatch(t *testing.T) {
 	}
 }
 
+func autoRemovalTestCleanup() {
+	instanceId := awsservice.GetInstanceId()
+	awsservice.DeleteLogGroupAndStream(instanceId, instanceId)
+	paths, _ := filepath.Glob(logFilePath + "*")
+	for _, p := range paths {
+		_ = os.Remove(p)
+	}
+}
+
+
+// TestAutoRemovalStopAgent configures agent to monitor a file with auto removal on.
+// Then it restarts the agent.
+// Verify the file is NOT removed.
+func TestAutoRemovalStopAgent(t *testing.T) {
+	configPath := "resources/config_auto_removal.json"
+	defer autoRemovalTestCleanup()
+
+	fpath := logFilePath + "1"
+	f, err := os.Create(fpath)
+	if err != nil {
+		t.Fatalf("Error occurred creating log file for writing: %v", err)
+	}
+	defer f.Close()
+
+	// Restart the agent multiple times.
+	loopCount := 5
+	linesPerLoop := 1000
+	start := time.Now()
+	for i := 0; i < loopCount; i++ {
+		common.StartAgent(configPath, true, false)
+		// Sleep to ensure agent detects file before it is written.
+		time.Sleep(sleepForFlush)
+		writeLogLines(t, f, linesPerLoop)
+		time.Sleep(sleepForFlush)
+		common.StopAgent()
+		assert.FileExists(t, fpath)
+	}
+
+	end := time.Now()
+	// Sleep to ensure backend stores logs.
+	time.Sleep(time.Second*60)
+	instanceId := awsservice.GetInstanceId()
+	err = awsservice.ValidateLogs(
+		instanceId,
+		instanceId,
+		&start,
+		&end,
+		// *2 because 2 lines per loop
+		awsservice.AssertLogsCount(loopCount * linesPerLoop * 2),
+		awsservice.AssertNoDuplicateLogs(),
+	)
+	assert.NoError(t, err)
+}
+
+// TestAutoRemovalFileRotation repeatedly created files matching the monitored pattern.
+// After creating each file, write some log lines, sleep and verify previous_file was auto removed.
+// Retrieve LogEvents from CWL and verify all log lines were uploaded.
+func TestAutoRemovalFileRotation(t *testing.T) {
```

Comment:
Is there a way to pull out shared code here?

Reply:
Yes always.
Not sure if it always improved readability, but I will create a few helper functions.